Custom aggregate functions with database translations fail during index selection

Custom aggregate functions that have database translations throw an exception.

java.lang.RuntimeException: Error while applying rule BindableAggregateRule(in:NONE,out:BINDABLE), args [rel#50963:LogicalAggregate.NONE.[](input=RelSubset#50962,group={29, 30},EXPR$2=jsonArrayAgg($31),__timestamp=MAX($32))]
	at org.apache.calcite.plan.volcano.VolcanoRuleCall.onMatch(VolcanoRuleCall.java:256)
	at org.apache.calcite.plan.volcano.IterativeRuleDriver.drive(IterativeRuleDriver.java:58)
	at org.apache.calcite.plan.volcano.VolcanoPlanner.findBestExp(VolcanoPlanner.java:510)
	at com.datasqrl.plan.rules.SQRLPrograms.lambda$static$0(SQRLPrograms.java:22)
	at com.datasqrl.plan.RelStageRunner.runStage(RelStageRunner.java:22)
	at com.datasqrl.plan.RelStageRunner.runStage(RelStageRunner.java:16)
	at com.datasqrl.plan.global.IndexSelector.getIndexSelection(IndexSelector.java:45)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at com.datasqrl.plan.global.DAGAssembler.assemble(DAGAssembler.java:168)
	at com.datasqrl.plan.global.DAGPlanner.assemble(DAGPlanner.java:86)
	at com.datasqrl.plan.global.DAGPlanner.plan(DAGPlanner.java:95)
	at com.datasqrl.frontend.SqrlOptimizeDag.planDag(SqrlOptimizeDag.java:50)
	at com.datasqrl.compile.Compiler.run(Compiler.java:172)
	at com.datasqrl.cmd.AbstractCompilerCommand.compilePackage(AbstractCompilerCommand.java:108)
	at com.datasqrl.cmd.AbstractCompilerCommand.runCommand(AbstractCompilerCommand.java:80)
	at com.datasqrl.cmd.AbstractCommand.run(AbstractCommand.java:30)
	at picocli.CommandLine.executeUserObject(CommandLine.java:2026)
	at picocli.CommandLine.access$1500(CommandLine.java:148)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2461)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2453)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2415)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2273)
	at picocli.CommandLine$RunLast.execute(CommandLine.java:2417)
	at picocli.CommandLine.execute(CommandLine.java:2170)
	at com.datasqrl.DatasqrlCMD.main(DatasqrlCMD.java:12)
Caused by: java.lang.IllegalStateException: User defined aggregation jsonArrayAgg must implement ImplementableAggFunction
	at org.apache.calcite.adapter.enumerable.RexImpTable.get(RexImpTable.java:778)
	at org.apache.calcite.interpreter.Bindables$BindableAggregate.<init>(Bindables.java:735)
	at org.apache.calcite.interpreter.Bindables$BindableAggregate.<init>(Bindables.java:748)
	at org.apache.calcite.interpreter.Bindables$BindableAggregateRule.convert(Bindables.java:800)
	at org.apache.calcite.rel.convert.ConverterRule.onMatch(ConverterRule.java:167)
	at org.apache.calcite.plan.volcano.VolcanoRuleCall.onMatch(VolcanoRuleCall.java:229)
	... 30 more
[FATAL] Error while applying rule BindableAggregateRule(in:NONE,out:BINDABLE), args [rel#50963:LogicalAggregate.NONE.[](input=RelSubset#50962,group={29, 30},EXPR$2=jsonArrayAgg($31),__timestamp=MAX($32))]

--- a/sqrl-planner/sqrl-common/src/main/java/com/datasqrl/flink/function/BridgingSqlAggregateFunction.java
+++ b/sqrl-planner/sqrl-common/src/main/java/com/datasqrl/flink/function/BridgingSqlAggregateFunction.java
@@ -20,10 +20,22 @@ package com.datasqrl.flink.function;
 
 import com.datasqrl.calcite.Dialect;
 import com.datasqrl.calcite.function.RuleTransform;
+import java.lang.reflect.Method;
+import java.lang.reflect.Type;
 import java.util.List;
 import lombok.Getter;
+import org.apache.calcite.adapter.enumerable.AggAddContext;
+import org.apache.calcite.adapter.enumerable.AggContext;
+import org.apache.calcite.adapter.enumerable.AggImplementor;
+import org.apache.calcite.adapter.enumerable.AggResetContext;
+import org.apache.calcite.adapter.enumerable.AggResultContext;
+import org.apache.calcite.linq4j.tree.Expression;
 import org.apache.calcite.plan.RelRule;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rel.type.RelDataTypeFactory;
 import org.apache.calcite.schema.AggregateFunction;
+import org.apache.calcite.schema.FunctionParameter;
+import org.apache.calcite.schema.ImplementableAggFunction;
 import org.apache.calcite.sql.SqlIdentifier;
 import org.apache.calcite.sql.SqlKind;
 import org.apache.calcite.sql.SqlOperator;
@@ -33,6 +45,7 @@ import org.apache.calcite.sql.type.SqlOperandTypeInference;
 import org.apache.calcite.sql.type.SqlReturnTypeInference;
 import org.apache.calcite.sql.validate.SqlUserDefinedAggFunction;
 import org.apache.calcite.util.Optionality;
+import org.apache.flink.calcite.shaded.com.google.common.collect.ImmutableList;
 import org.apache.flink.table.catalog.DataTypeFactory;
 import org.apache.flink.table.functions.FunctionDefinition;
 import org.apache.flink.table.planner.calcite.FlinkTypeFactory;
@@ -111,25 +124,45 @@ public class BridgingSqlAggregateFunction extends SqlUserDefinedAggFunction impl
 //        valueTypes.add(type);
 //      }
 //    }
-//
-//    return new ImplementableAggFunction() {
-//      @Override
-//      public RelDataType getReturnType(RelDataTypeFactory relDataTypeFactory) {
-//        throw new RuntimeException("todo");
-//      }
-//
-//      @Override
-//      public AggImplementor getImplementor(boolean b) {
-//        return new RexImpTable.UserDefinedAggReflectiveImplementor(this);
-//      }
-//
-//      @Override
-//      public List<FunctionParameter> getParameters() {
-//        //derive parameters (necessary?)
-//        return List.of();
-//      }
-//    };
-    return null;
+
+    return new ImplementableAggFunction() {
+
+      @Override
+      public List<FunctionParameter> getParameters() {
+        return List.of();
+      }
+
+      @Override
+      public RelDataType getReturnType(RelDataTypeFactory relDataTypeFactory) {
+        return null;
+      }
+
+      @Override
+      public AggImplementor getImplementor(boolean b) {
+        return new AggImplementor() {
+          @Override
+          public List<Type> getStateType(AggContext aggContext) {
+            return null;
+          }
+
+          @Override
+          public void implementReset(AggContext aggContext, AggResetContext aggResetContext) {
+
+          }
+
+          @Override
+          public void implementAdd(AggContext aggContext, AggAddContext aggAddContext) {
+
+          }
+
+          @Override
+          public Expression implementResult(AggContext aggContext,
+              AggResultContext aggResultContext) {
+            return null;
+          }
+        };
+      }
+    };
   }
 
   @Override

--- a/sqrl-testing/sqrl-integration-tests/src/test/java/com/datasqrl/flink/FlinkPhysicalIT.java
+++ b/sqrl-testing/sqrl-integration-tests/src/test/java/com/datasqrl/flink/FlinkPhysicalIT.java
@@ -74,6 +74,15 @@ class FlinkPhysicalIT extends AbstractPhysicalSQRLIT {
   }
 
   @Test
+  public void jsonAggTest() {
+    ScriptBuilder builder = example.getImports();
+    builder.add("IMPORT json.*;");
+    builder.add("BigJoin := SELECT * FROM Customer JOIN Product on true;");
+    builder.add("Agg := SELECT jsonArrayAgg(name) AS agg FROM BigJoin GROUP BY name;");
+    validateTables(builder.getScript(), "Agg");
+  }
+
+  @Test
   public void functionTest() {
     ScriptBuilder builder = example.getImports();
     builder.add("IMPORT text.*");

--- a/sqrl-testing/sqrl-integration-tests/src/test/resources/snapshots/com/datasqrl/flink/FlinkPhysicalIT/jsonAggTest.txt
+++ b/sqrl-testing/sqrl-integration-tests/src/test/resources/snapshots/com/datasqrl/flink/FlinkPhysicalIT/jsonAggTest.txt
@@ -1,0 +1,21 @@
+>>>customer$2-lp-stream
+LogicalTableScan(table=[[customer$1]], hints=[[[WatermarkHint inheritPath:[] options:[1]], [TimestampHint inheritPath:[] options:[1]]]]) hints[WatermarkHint options:[1], TimestampHint options:[1]]
+
+>>>product$2-lp-stream
+LogicalTableScan(table=[[product$1]], hints=[[[WatermarkHint inheritPath:[] options:[1]], [TimestampHint inheritPath:[] options:[1]]]]) hints[WatermarkHint options:[1], TimestampHint options:[1]]
+
+>>>Agg-lp-database
+LogicalSort(sort0=[$0], dir0=[ASC])
+  LogicalProject(_name=[$0], agg=[$1], __timestamp=[$2])
+    LogicalProject(_name=[$0], agg=[$1], __timestamp=[$2])
+      LogicalAggregate(group=[{5}], agg=[jsonArrayAgg($5)], __timestamp=[MAX($12)])
+        LogicalProject(_uuid=[$0], _uuid0=[$6], _ingest_time=[$1], customerid=[$2], email=[$3], name=[$4], lastUpdated=[$5], _ingest_time0=[$7], productid=[$8], name0=[$9], description=[$10], category=[$11], __timestamp=[CASE(<($1, $7), $7, $1)])
+          LogicalJoin(condition=[true], joinType=[inner]) hints[JoinCostHint options:[STREAM, STREAM, 0, NONE]]
+            LogicalTableScan(table=[[customer$2]])
+            LogicalTableScan(table=[[product$2]])
+
+>>>Agg-data
+Emily F. Ludies, ["Emily F. Ludies", "Emily F. Ludies", "Emily F. Ludies", "Emily F. Ludies", "Emily F. Ludies", "Emily F. Ludies"]
+John Mekker, ["John Mekker", "John Mekker", "John Mekker", "John Mekker", "John Mekker", "John Mekker"]
+Mark Spielman, ["Mark Spielman", "Mark Spielman", "Mark Spielman", "Mark Spielman", "Mark Spielman", "Mark Spielman"]
+Michelle Dagnes, ["Michelle Dagnes", "Michelle Dagnes", "Michelle Dagnes", "Michelle Dagnes", "Michelle Dagnes", "Michelle Dagnes"]

